addons/utm/models/utm_source.py

```diff
@@ -71,13 +71,38 @@ def create(self, vals_list):
 
         return super().create(vals_list)
 
-    def write(self, values):
-        if values.get(self._rec_name) and not values.get('name'):
-            values['name'] = self.env['utm.source']._generate_name(self, values[self._rec_name])
-        if values.get('name'):
-            values['name'] = self.env['utm.mixin']._get_unique_names(self._name, [values['name']])[0]
+    def find_smalles_missing(self, values):
+        num_set = set(values)
+        current = 2
+        while True:
+            if current not in num_set:
+                return current
+            current += 1
 
-        super().write(values)
+    def write(self, values):
+        mixin = self.env['utm.mixin']
+        name = values.get('name')
+        if values.get(self._rec_name) and not name:
+            name = self.env['utm.source']._generate_name(self, values[self._rec_name])
+
+        if name:
+            name_without_counter, _ = mixin._split_name_and_count(name)
+            similar_names = self.search([('name', 'ilike', name_without_counter)])
+            similar_names = similar_names.filtered(lambda r: mixin._split_name_and_count(r.name)[0] == name_without_counter)
+            similar_names_editing = self & similar_names
+            need_update = self - similar_names_editing
+            if len(need_update) >= 2:
+                raise ValueError(_('Multiple records with the same name. The name should be unique!'))\
+
+            if name_without_counter:
+                super(UtmSourceMixin, similar_names_editing).write(values)
+
+            smallest_num = self.find_smalles_missing([mixin._split_name_and_count(name)[1] for name in similar_names.mapped('name')])
+            if name:
+                values['name'] = f'{name} [{smallest_num}]' if smallest_num > 2 else name
+            super(UtmSourceMixin, need_update).write(values)
+        else:
+            return super().write(values)
```

Comment:
So, just to make sure, can't we get away with a simple change like this for this issue?
```suggestion
    def write(self, values):
        if values.get(self._rec_name) and not values.get('name'):
            values['name'] = self.env['utm.source']._generate_name(self, values[self._rec_name])
        if name and len(self) > 1:
            raise ValueError(_('Multiple records with the same name. The name should be unique!'))
        if values.get('name') and  mixin._split_name_and_count(name)[0] == mixin._split_name_and_count(self[0].name)[0]:
            values['name'] = self.env['utm.mixin']._get_unique_names(self._name, [values['name']])[0]
```